File: backend/fastapi/app/services/tasksync.py

import base64
from datetime import datetime
from uuid import uuid4

def _ics_event(dtstart_date: str, summary: str, uid: str | None = None, description: str = ""):
    """Generate a minimal ICS ``VEVENT`` block."""
    uid = uid or str(uuid4())
    return (
        f"BEGIN:VEVENT\n"
        f"UID:{uid}\n"
        f"DTSTAMP:{datetime.utcnow().strftime('%Y%m%dT%H%M%SZ')}\n"
        f"DTSTART;VALUE=DATE:{dtstart_date.replace('-', '')}\n"
        f"SUMMARY:{summary}\n"
        f"DESCRIPTION:{description}\nEND:VEVENT"
    )

def create_task_sync(payload: dict):
    """Create calendar events from a timeline and return an ICS file."""
    tl = payload.get("timeline", {})
    events = []
    if tl.get("determinationDue"):
        events.append(_ics_event(tl["determinationDue"], "CPRA Determination Due"))
    if tl.get("extensionDue"):
        events.append(_ics_event(tl["extensionDue"], "CPRA Extension Due"))
    for m in tl.get("milestones", []):
        if m.get("label") == "Draft production":
            events.append(_ics_event(m["due"], "CPRA Draft Production"))
    ics = "BEGIN:VCALENDAR\nVERSION:2.0\nPRODID:-//CPRA Planner//EN\n" + "\n".join(events) + "\nEND:VCALENDAR\n"
    return {
        "createdEvents": [{"provider": "ics", "id": "local", "url": ""}],
        "icsFileBase64": base64.b64encode(ics.encode()).decode(),
        "emailsSent": 0,
    }

File: backend/express/src/services/tasksync.ts
import { randomUUID } from 'crypto';

function icsEvent(dtstart_date: string, summary: string, uid?: string, description = ''): string {
  const id = uid || randomUUID();
  const dtstamp = new Date().toISOString().replace(/[-:]/g, '').split('.')[0] + 'Z';
  return `BEGIN:VEVENT\nUID:${id}\nDTSTAMP:${dtstamp}\nDTSTART;VALUE=DATE:${dtstart_date.replace(/-/g,'')}\nSUMMARY:${summary}\nDESCRIPTION:${description}\nEND:VEVENT`;
}

export function createTaskSync(payload: any) {
  const tl = payload.timeline || {};
  const events: string[] = [];
  if (tl.determinationDue) events.push(icsEvent(tl.determinationDue, 'CPRA Determination Due'));
  if (tl.extensionDue) events.push(icsEvent(tl.extensionDue, 'CPRA Extension Due'));
  for (const m of tl.milestones || []) {
    if (m.label === 'Draft production') events.push(icsEvent(m.due, 'CPRA Draft Production'));
  }
  const ics = `BEGIN:VCALENDAR\nVERSION:2.0\nPRODID:-//CPRA Planner//EN\n${events.join('\n')}\nEND:VCALENDAR\n`;
  return {
    createdEvents: [{ provider: 'ics', id: 'local', url: '' }],
    icsFileBase64: Buffer.from(ics).toString('base64'),
    emailsSent: 0,
  };
}
